libraries/anvill_passes/src/ConvertXorToCmp.cpp

```diff
@@ -0,0 +1,198 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <anvill/ABI.h>
+#include <glog/logging.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/InstIterator.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Pass.h>
+
+#include <vector>
+
+namespace anvill {
+namespace {
+
+class ConvertXorToCmp final : public llvm::FunctionPass {
+ public:
+  ConvertXorToCmp(void) : llvm::FunctionPass(ID) {}
+
+  bool runOnFunction(llvm::Function &func) override;
+
+ private:
+  static char ID;
+};
+
+char ConvertXorToCmp::ID = '\0';
+
+// Get which operand for a binary operator is an ICmpInst
+// nullptr if neither
+static llvm::ICmpInst *getComparisonOperand(llvm::BinaryOperator *op) {
+  auto rhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(1));
+  if (rhs_cmp) {
+    return rhs_cmp;
+  }
+
+  auto lhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(op->getOperand(0));
+  if (lhs_cmp) {
+    return lhs_cmp;
+  }
+
+  return nullptr;
+}
+
+static llvm::Value *negateCmpPredicate(llvm::ICmpInst *cmp) {
+  auto pred = cmp->getPredicate();
+  llvm::IRBuilder<> ir(cmp);
+  llvm::ICmpInst::Predicate new_pred = llvm::CmpInst::getInversePredicate(pred);
+
+  // create a new compare with negated predicate
+  return ir.CreateICmp(new_pred, cmp->getOperand(0), cmp->getOperand(1));
+}
+
+bool ConvertXorToCmp::runOnFunction(llvm::Function &func) {
+  std::vector<llvm::BinaryOperator *> xors;
+
+  for (auto &inst : llvm::instructions(func)) {
+
+    // check for binary op
+    if (auto binop = llvm::dyn_cast<llvm::BinaryOperator>(&inst)) {
+
+      // binary op is a xor
+      if (binop->getOpcode() == llvm::Instruction::Xor) {
+
+        // xor has a constant as either lhs or rhs, and its a one-bit constant of size 1
+        auto lhs_c = llvm::dyn_cast<llvm::ConstantInt>(binop->getOperand(0));
+        auto lhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(binop->getOperand(0));
+
+        auto rhs_c = llvm::dyn_cast<llvm::ConstantInt>(binop->getOperand(1));
+        auto rhs_cmp = llvm::dyn_cast<llvm::ICmpInst>(binop->getOperand(1));
+
+
+        // left side is a constant int 1 (true), right side is a cmpinst
+        if (lhs_c && lhs_c->getType()->getBitWidth() == 1 &&
+            lhs_c->isAllOnesValue() && rhs_cmp) {
+          xors.emplace_back(binop);
+        } else if (rhs_c && rhs_c->getType()->getBitWidth() == 1 &&
+            rhs_c->isAllOnesValue() && lhs_cmp) {
+          // right side is a constant int 1 (true), left side is a cmp
+          xors.emplace_back(binop);
+        }
+      }
+    }
+  }
+
+  auto changed = false;
+  int replaced_items = 0;
+
+  for (auto xori : xors) {
+
+    // find predicate from xor's operands
+    auto cmp = getComparisonOperand(xori);
+    if (!cmp) {
+      DLOG(ERROR)
+          << "Error: Xor should have a comparison operand, but we can't find it again";
+      continue;
+    }
+
+    std::vector<llvm::BranchInst *> brs_to_invert;
```

Comment:
Move these vectors outside of your `for` loop, and clear them inside so that you can re-use the backing storage across iterations of `xori`.